File: android/app/src/main/java/com/openburnbar/data/policy/MobileSyncOwnershipPolicy.kt
package com.openburnbar.data.policy

enum class MobileSyncPublisherRole(val wire: String) {
    MAC_PUBLISHES("mac-publishes"),
    MOBILE_MIRRORS_READ_ONLY("mobile-mirrors-read-only"),
}

enum class MobileSyncFreshness(val wire: String, val label: String) {
    LIVE("live", "Live"),
    STALE("stale", "Stale"),
    OFFLINE("offline", "Offline"),
    EMPTY("empty", "No Mac-published data"),
    FAILED("failed", "Cloud load failed"),
    PARTIAL("partial", "Partial"),
    ;

    val looksLikeLiveZero: Boolean get() = this == LIVE
}

object MobileSyncOwnershipPolicy {
    val macRole = MobileSyncPublisherRole.MAC_PUBLISHES
    val mobileRole = MobileSyncPublisherRole.MOBILE_MIRRORS_READ_ONLY
    const val mobileMayPublishUsage = false

    fun freshness(
        hasData: Boolean,
        failed: Boolean,
        offline: Boolean,
        stale: Boolean,
        partial: Boolean,
    ): MobileSyncFreshness = when {
        failed -> MobileSyncFreshness.FAILED
        offline -> MobileSyncFreshness.OFFLINE
        !hasData -> MobileSyncFreshness.EMPTY
        partial -> MobileSyncFreshness.PARTIAL
        stale -> MobileSyncFreshness.STALE
        else -> MobileSyncFreshness.LIVE
    }

    fun shouldApply(startedGeneration: Int, currentGeneration: Int, cancelled: Boolean): Boolean =
        !cancelled && startedGeneration == currentGeneration

    fun nextGeneration(current: Int): Int = current + 1
}

File: OpenBurnBarCore/Sources/OpenBurnBarMobilePolicy/MobileSyncOwnershipPolicy.swift
import Foundation

/// Mac publishes; mobile mirrors. Mobile never becomes the write owner of
/// usage/rollup/quota snapshots.
public enum MobileSyncPublisherRole: String, Sendable, Equatable {
    case macPublishes = "mac-publishes"
    case mobileMirrorsReadOnly = "mobile-mirrors-read-only"
}

public enum MobileSyncFreshness: String, Sendable, Equatable, CaseIterable {
    case live
    case stale
    case offline
    case empty
    case failed
    case partial

    /// Empty/failed/offline must never render as a live zero.
    public var looksLikeLiveZero: Bool { self == .live }

    public var label: String {
        switch self {
        case .live: return "Live"
        case .stale: return "Stale"
        case .offline: return "Offline"
        case .empty: return "No Mac-published data"
        case .failed: return "Cloud load failed"
        case .partial: return "Partial"
        }
    }
}

public enum MobileSyncOwnershipPolicy {
    public static let macRole = MobileSyncPublisherRole.macPublishes
    public static let mobileRole = MobileSyncPublisherRole.mobileMirrorsReadOnly

    public static var mobileMayPublishUsage: Bool { false }

    public static func freshness(
        hasData: Bool,
        failed: Bool,
        offline: Bool,
        stale: Bool,
        partial: Bool
    ) -> MobileSyncFreshness {
        if failed { return .failed }
        if offline { return .offline }
        if !hasData { return .empty }
        if partial { return .partial }
        if stale { return .stale }
        return .live
    }

    /// Cancel and process-restart share one generation token so a late result
    /// cannot double-apply.
    public static func shouldApply(
        startedGeneration: Int,
        currentGeneration: Int,
        cancelled: Bool
    ) -> Bool {
        !cancelled && startedGeneration == currentGeneration
    }

    public static func nextGeneration(_ current: Int) -> Int {
        current &+ 1
    }
}
